text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+- Feature Name: `const_trait_methods`
+- Start Date: 2024-12-13
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#67792](https://github.com/rust-lang/rust/issues/67792)
+
+# Summary
+[summary]: #summary
+
+Make trait methods callable in const contexts. This includes the following parts:
+
+* Allow marking `trait` declarations as const implementable.
+* Allow marking `trait` impls as `const`.
+* Allow marking trait bounds as `const` to make methods of them callable in const contexts.
+
+Fully contained example ([Playground of currently working example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2ab8d572c63bcf116b93c632705ddc1b)):
+
+```rust
+trait Default {
+    (const) fn default() -> Self;
+}
```

Comment:
It's important to highlight somewhere that the intended intuition behind the parentheses in this position is that what's in them is tied to the associated effect.  Following from this, @nikomatsakis and I mean to allow this as well:

```rust
impl Default for () {
    () fn default() {}
}
```

And for that matter:

```rust
trait Default {
    () fn default() -> Self;
}

These have the same semantics as if the `()` wasn't there.

We're not suggesting that we're necessarily going to encourage this, simply that we want to allow it for consistency and pedagogy.  When teaching people, using this form is part of our plan for helping people to really see what's going on.

So anyway, we should mention somewhere in the RFC that this syntax is allowed.

Reply:
We could lint against it, but disable the lint in macros. Then ppl can write macros that pick const or not easily

Reply:
Good point about it being good for macros.

Reply:
This syntax has problems for future `(const) fn()` pointers, as these are types the parser will parse `() fn()` as unit followed by more characters. I'm just not gonna support this and we can discuss it in the future